### PR TITLE
Refactor level checking / type healing logic

### PR DESCRIPTION
**Pull request by nicolasstucki**
Disentangle level checking from type healing.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/Splicing.scala [hunk 1]
```diff
@@ -77,7 +77,7 @@ class Splicing extends MacroTransform:
 
   override def run(using Context): Unit =
     if ctx.compilationUnit.needsStaging then
-      super.run(using freshStagingContext)
+      super.run
```
Comment (natsukagami) [inline compiler/src/dotty/tools/dotc/transform/Splicing.scala:80]: Why are some of the calls have `freshStagingContext` removed, while some are replaced with `StagingLevel.freshContext`? I don't really understand why or when which is used.

Comment (natsukagami) [reply compiler/src/dotty/tools/dotc/transform/Splicing.scala:80]: Oh it seems like `StagingLevel.freshContext` are used in the tests...?

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/Splicing.scala:80]: I changed some instances in `tests/pos-with-compiler-cc/dotc` by mistake. I reverted those changes. That test should not change.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/Splicing.scala:80]: Note that I do not initialize the levels until we add the first symbol with `level != 0`.

File: compiler/src/dotty/tools/dotc/staging/HealType.scala [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+  private def dealiasAndTryHeal(sym: Symbol, tp: TypeRef, pos: SrcPos): Type =
+    val tp1 = tp.dealias
+    if tp1 != tp then apply(tp1)
+    else tryHeal(tp.symbol, tp, pos)
+
+  /** Try to heal reference to type `T` used in a higher level than its definition.
+   *  Returns a reference to a type tag generated by `QuoteTypeTags` that contains a
+   *  reference to a type alias containing the equivalent of `${summon[quoted.Type[T]]}`.
+   *  Emits and error if `T` cannot be healed and returns `T`.
```
Comment (natsukagami) [inline compiler/src/dotty/tools/dotc/staging/HealType.scala:69]:
```suggestion
   *  Emits an error if `T` cannot be healed and returns `T`.
```

**Review by natsukagami — APPROVED** [review]

**Pull request merged**